### PR TITLE
Rework quality measure approach

### LINKED ISSUE #88: Add error messages in `quality.py`
The quality metrics in `quality.py` often raise bare `NotImplementedError`s. It would be good to provide clearer error reporting and perhaps to combine the metric implementations to avoid duplication.

### PR DESCRIPTION
**Pull request by jwallwork23**
Closes #88.

Reworks Pyroteus' quality measures into a class to cut down on duplicated code.

@nimberledge Hello! If you have some spare time and are interested then it would be great to hear your perspective on this. If you don't have time then don't worry about it :)

### CODE REVIEW

File: test/test_quality.py [hunk 1]
```diff
@@ -31,103 +32,109 @@ class TestQuality(unittest.TestCase):
...
-        if name.startswith("get_quality_metrics"):
+        if name == "metric":
             P1_ten = TensorFunctionSpace(mesh, "CG", 1)
             M = interpolate(Identity(dim), P1_ten)
-            return measure(mesh, M, **kwargs)
-        return measure(mesh, **kwargs)
+            qm = QualityMeasure(mesh, metric=M, **kwargs)
+        else:
+            qm = QualityMeasure(mesh, **kwargs)
+        return qm(name)
```
Comment (jwallwork23) [inline test/test_quality.py:41]: The following would be more concise:
```
    kwargs["metric"] = M
return QualityMeasure(mesh, **kwargs)(name)
```

File: test/test_quality.py [hunk 2]
```diff
@@ -31,103 +32,109 @@ class TestQuality(unittest.TestCase):
...
     )
     def test_uniform_quality_2d(self, measure, expected):
         mesh = uniform_mesh(2, 10)
         q = self.quality(measure, mesh)
         truth = Function(q.function_space()).assign(expected)
         self.assertAlmostEqual(errornorm(truth, q), 0.0, places=6)
-        if measure == "get_areas2d":
+        if measure == "areas2d":
             s = q.vector().gather().sum()
             self.assertAlmostEqual(s, 1.0)
```
Comment (jwallwork23) [inline test/test_quality.py:61]: Oops this never happens

File: test/test_quality.py [hunk 3]
```diff
@@ -31,103 +32,109 @@ class TestQuality(unittest.TestCase):
...
     )
     def test_uniform_quality_3d(self, measure, expected):
         mesh = uniform_mesh(3, 4)
         q = self.quality(measure, mesh)
         truth = Function(q.function_space()).assign(expected)
         self.assertAlmostEqual(errornorm(truth, q), 0.0)
-        if measure == "get_volumes3d":
+        if measure == "volumes3d":
             s = q.vector().gather().sum()
             self.assertAlmostEqual(s, 1.0)
```
Comment (jwallwork23) [inline test/test_quality.py:80]: Nor this.

File: pyroteus/quality.py [hunk 4]
```diff
@@ -15,463 +15,134 @@
...
+        "metric",
+    )
+
+    @PETSc.Log.EventDecorator()
+    def __init__(
+        self, mesh: MeshGeometry, metric: Function = None, python: bool = False
+    ):
+        """
+        :arg mesh: the input mesh to do computations on
+        :arg metric: the tenor field representing the metric space transformation
```
Comment (nimberledge) [inline pyroteus/quality.py:54]: typo in word "tensor"

**Review by nimberledge — COMMENTED** [review]
Much cleaner this way, I like it.

File: pyroteus/quality.py (hunk 4 above)
Comment (jwallwork23) [reply pyroteus/quality.py]: Good spot. Addressed in https://github.com/pyroteus/pyroteus/pull/99/commits/8da049a3db40acea91225ce0771da38e9057d4e6.

**Comment by jwallwork23** [comment]
Thanks very much for taking the time to look at this @nimberledge! Please could you click "approve" if happy?

**Review by nimberledge — APPROVED** [review]

**Pull request merged**